xarray/tests/test_weighted.py

```diff
@@ -0,0 +1,324 @@
+def test_weighted__sum_of_weights_2D(shape, dim):
+
+    np.random.seed(0)
+
+    data = np.zeros(shape)
+    # make sure all weights are positive to avoid summing to 0
+    weights = np.abs(np.random.randn(*shape))
+
+    da = DataArray(data)
+    weights = DataArray(weights)
+
+    weighted = da.weighted(weights)
+
+    expected = weights.sum(dim=dim)
+    result = weighted.sum_of_weights(dim=dim)
+
+    assert_allclose(expected, result)
+
+
+def test_weigted_non_DataArray_weights():
+
+    da = DataArray([1, 2])
+    with raises_regex(AssertionError, "'weights' must be a DataArray"):
+        da.weighted([1, 2])
+
+    ds = Dataset(dict(data=[1, 2]))
+    with raises_regex(AssertionError, "'weights' must be a DataArray"):
+        ds.weighted([1, 2])
+
+
+@pytest.mark.parametrize("weights", ([np.nan, 2], [np.nan, np.nan]))
+def test_weighted_weights_nan_raises(weights):
+    # make sure NaNs in weights raise
+
+    with pytest.raises(ValueError, match="`weights` cannot contain missing values."):
+        DataArray([1, 2]).weighted(DataArray(weights))
+
+
+@pytest.mark.parametrize(
+    ("weights", "expected"),
+    (([1, 2], 3), ([0, 2], 2), ([0, 0], np.nan), ([-1, 1], np.nan)),
+)
+def test_weigted_sum_of_weights_no_nan(weights, expected):
+
+    da = DataArray([1, 2])
+    weights = DataArray(weights)
+    result = da.weighted(weights).sum_of_weights()
+
+    expected = DataArray(expected)
+
+    assert_equal(expected, result)
+
+
+@pytest.mark.parametrize(
+    ("weights", "expected"),
+    (([1, 2], 2), ([0, 2], 2), ([0, 0], np.nan), ([-1, 1], 1), ([2, 0], np.nan)),
+)
+def test_weigted_sum_of_weights_nan(weights, expected):
+
+    da = DataArray([np.nan, 2])
+    weights = DataArray(weights)
+    result = da.weighted(weights).sum_of_weights()
+
+    expected = DataArray(expected)
+
+    assert_equal(expected, result)
+
+
+@pytest.mark.parametrize("da", ([1, 2], [1, np.nan], [np.nan, np.nan]))
+@pytest.mark.parametrize("factor", [0, 1, 2, 3.14])
+@pytest.mark.parametrize("skipna", (True, False))
+def test_weighted_sum_equal_weights(da, factor, skipna):
+    # if all weights are 'f'; weighted sum is f times the ordinary sum
+
+    da = DataArray(da)
+    weights = xr.zeros_like(da) + factor
+
+    expected = da.sum(skipna=skipna) * factor
+    result = da.weighted(weights).sum(skipna=skipna)
+
+    assert_equal(expected, result)
+
+
+@pytest.mark.parametrize(
+    ("weights", "expected"), (([1, 2], 5), ([0, 2], 4), ([0, 0], 0))
+)
+def test_weighted_sum_no_nan(weights, expected):
+    da = DataArray([1, 2])
+
+    weights = DataArray(weights)
+    result = da.weighted(weights).sum()
+    expected = DataArray(expected)
+
+    assert_equal(expected, result)
+
+
+@pytest.mark.parametrize(
+    ("weights", "expected"), (([1, 2], 4), ([0, 2], 4), ([1, 0], 0), ([0, 0], 0))
+)
+@pytest.mark.parametrize("skipna", (True, False))
+def test_weighted_sum_nan(weights, expected, skipna):
```

Comment:
can these tests be removed in favour of `test_weighted_operations_different_shapes` and test_weighted_operations_3D`?

Reply:
I'll go through them again - I want to keep some "manual" tests

Reply:
I removed some tests but kept most of them - I like to have some `manual` tests and some of these test edge cases (e.g. `DataArray([np.nan, 1]).weighted(DataArray([1, 0])).mean()` (`weights` sum to zero because there are `NaNs` in the data)).